src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -648,28 +689,58 @@ internal override int Read(Span<byte> buffer)
         {
             ThrowIfDisposed();
             byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length);
+            CancellationTokenSource? cts = null;
             try
             {
-                Task<int> t = ReadAsync(new Memory<byte>(rentedBuffer, 0, buffer.Length)).AsTask();
+                if (_readTimeout > 0)
+                {
+                    cts = new CancellationTokenSource(_readTimeout);
+                }
+                Task<int> t = ReadAsync(new Memory<byte>(rentedBuffer, 0, buffer.Length), cts != null ? cts.Token : default).AsTask();
                 ((IAsyncResult)t).AsyncWaitHandle.WaitOne();
                 int readLength = t.GetAwaiter().GetResult();
                 rentedBuffer.AsSpan(0, readLength).CopyTo(buffer);
                 return readLength;
             }
+            catch (OperationCanceledException) when (_readTimeout > 0)
```

Comment:
```suggestion
            catch (OperationCanceledException) when (cts != null && cts.IsCancellationRequested)
```
?

Reply:
I like it as it is more specific to the Token we create. updated.

Reply:
Shouldn't we specifically check for the token on the Exception?

```csharp
    catch (OperationCanceledException oce) when (oce.Token == cts?.Token)
```

@stephentoub?